tests/zquantum/optimizers/layerwise_ansatz_optimizer_test.py

```diff
@@ -4,98 +4,86 @@
+from zquantum.core.interfaces.functions import FunctionWithGradient
 
 import pytest
 from functools import partial
 
 
+ansatz = MockAnsatz(1, 5)
+
+
 @pytest.fixture(
     params=[
-        {"inner_optimizer": ScipyOptimizer("L-BFGS-B"), "min_layer": 1, "max_layer": 1}
+        {
+            "ansatz": ansatz,
+            "inner_optimizer": ScipyOptimizer("L-BFGS-B"),
+            "min_layer": 1,
+            "max_layer": 3,
+        }
     ]
 )
 def optimizer(request):
     return LayerwiseAnsatzOptimizer(**request.param)
 
 
-@pytest.fixture(params=[True, False])
-def keep_history(request):
-    return request.param
-
+def cost_function_factory(ansatz):
+    def cost_function(x):
+        return sum(x ** 2) * ansatz.number_of_layers
 
-class TestLayerwiseAnsatzOptimizer(OptimizerTests):
-    @pytest.fixture
-    def sum_x_squared(self):
-        class my_fun:
-            def __init__(self, ansatz):
-                self.ansatz = ansatz
+    return cost_function
 
-            def __call__(self, x):
-                return sum(x ** 2)
 
-        return my_fun(MockAnsatz(1, 5))
+initial_params = np.array([1])
 
-    @pytest.fixture
-    def rosenbrock_function(self):
-        def _rosenbrock_function(x):
-            return sum(100.0 * (x[1:] - x[:-1] ** 2.0) ** 2.0 + (1 - x[:-1]) ** 2.0)
 
-        _rosenbrock_function.ansatz = MockAnsatz(1, 5)
-
-        return _rosenbrock_function
-
-    def test_gradients_history_is_recorded_if_keep_history_is_true(
-        self, optimizer, sum_x_squared
-    ):
-        pytest.xfail(
-            """This test fail since LayerwiseAnsatzOptimizer is deepcopying 
-                input cost_function, which is intended behaviour."""
-        )
+class TestLayerwiseAnsatzOptimizer:
+    @pytest.mark.parametrize("contract", NESTED_OPTIMIZER_CONTRACTS)
+    def test_if_satisfies_contracts(self, contract, optimizer):
+        assert contract(optimizer, cost_function_factory, initial_params)
 
-    def test_ansatz_is_not_modified_outside_of_minimize(self, sum_x_squared):
-        optimizer = LayerwiseAnsatzOptimizer(
-            inner_optimizer=ScipyOptimizer("L-BFGS-B"), min_layer=2, max_layer=4
-        )
-        cost_function = sum_x_squared
-        initial_number_of_layers = cost_function.ansatz.number_of_layers
-        _ = optimizer.minimize(cost_function, initial_params=[0, 0])
-        assert cost_function.ansatz.number_of_layers == initial_number_of_layers
+    def test_ansatz_is_not_modified_outside_of_minimize(self, optimizer):
+        initial_number_of_layers = ansatz.number_of_layers
+        _ = optimizer.minimize(cost_function_factory, initial_params=initial_params)
+        assert ansatz.number_of_layers == initial_number_of_layers
 
     @pytest.mark.parametrize("max_layer", [2, 3, 4, 5])
-    def test_length_of_parameters_in_history_increases(self, sum_x_squared, max_layer):
+    def test_length_of_parameters_in_history_increases(self, max_layer):
```

Comment:
I might be wrong, but it seems to me that the description of the test does not match the actual test. We are not looking at history, but only at the dimension of the obtained optimal parameters.